LUCENE-10197: UnifiedHighlighter should use builders for thread-safety

apanimesh061 opened the pull request:
# Description

`UnifiedHighlighter` is not thread-safe due to the presence of setters. We can move the fields to builder so that the class becomes thread-safe. Objective is to made `UnifiedHighlighter` immutable.

# Solution

Add a builder that is able to set all fields of the `UnifiedHighlighter` class.

# Tests

```
./gradlew -p lucene/highlighter test
./gradlew check
```

# Checklist

Please review the following and check all that apply:

- [X] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [X] I have created a Jira issue and added the issue ID to my pull request title.
- [X] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [X ] I have developed this patch against the `main` branch.
- [X] I have run `./gradlew check`.
- [X] I have added tests for my changes.

apanimesh061 committed:
LUCENE-10197

UnifiedHighlighter is not thread-safe due to the presence of setters. We can move the fields to builder so that the class becomes thread-safe.

apanimesh061 committed:
gradlew check fixes

apanimesh061 committed:
Making builder extensible

apanimesh061 committed:
precommit lint fixes

apanimesh061 committed:
Adding randomUnifiedHighlighterWithBuilder to TestUnifiedHighlighter

apanimesh061 committed:
Move highlight flag set to UH class

apanimesh061 committed:
Address comments on the PR

apanimesh061 commented:
The new changes include the following:

1. if `flags` is initialized then boolean flags are ignored
2. Unit test for new `getFlags()` implementation using builders
3. Setter for `flags`
4. Builder can used to set booleans as well as `flags`

Have not implemented the randomized testing as suggested. Still researching into how to implement that.

dsmiley commented:
> Have not implemented the randomized testing as suggested.

I didn't mean for you to do any new testing.  Randomized testing of flags is already happening.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -182,6 +313,10 @@ public void setMaxLength(int maxLength) {
     this.maxLength = maxLength;
   }
 
+  public void setFlags(Set<HighlightFlag> flags) {
```
Uh; no, UH is now mutable again.  I'm dubious we even want this on the builder since it's more of an advanced case to customize these flags.  Someone can always subclass the getter to customize.

apanimesh061 replied:
Yes. I was working in direction of creating the builder only and later modifying the tests after the setters are removed. I will now do both the things: remove setters making UH immutable and modify unit tests to use the builder instead of the earlier setup.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -165,6 +275,27 @@ public UnifiedHighlighter(IndexSearcher indexSearcher, Analyzer indexAnalyzer) {
+  public UnifiedHighlighter(Builder<?> builder) {
+    this.searcher = builder.searcher;
+    this.indexAnalyzer = Objects.requireNonNull(builder.indexAnalyzer);
+    this.flags = builder.flags;
```
I propose a builder.getFlags() so you can compute the appropriate flags there, and someone could even override the builder there if they choose.

apanimesh061 replied:
Okay, I finally understand what I was misunderstanding. The builder getFlags will evaluate the flags based on booleans and the outer class's getFlags will just return the set like a regular getter.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -165,6 +275,27 @@ public UnifiedHighlighter(IndexSearcher indexSearcher, Analyzer indexAnalyzer) {
+    this.flags = builder.flags;
+    this.defaultHandleMtq = builder.handleMultiTermQuery;
+    this.defaultHighlightPhrasesStrictly = builder.highlightPhrasesStrictly;
+    this.defaultPassageRelevancyOverSpeed = builder.passageRelevancyOverSpeed;
```
no; I don't think we want all these booleans on UnifiedHighlighter when we have the flags.  The booleans can stay on the builder and inform how the flags there is computed.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -207,18 +342,22 @@ public void setFieldMatcher(Predicate<String> predicate) {
+   * Returns the predicate to use for extracting the query part that must be highlighted. By default
+   * only queries that target the current field are kept. (AKA requireFieldMatch)
    */
+  protected Predicate<String> getFieldMatcher(String field) {
```
assuming we move the default selection logic to the builder, this can be implemented like a getter; no?

apanimesh061 replied:
Yes you are right but ideally the getter should be `getFieldMatcher()` and not depend on the `field` argument? In case the input value was null, the output depends on the `field` argument (`(qf) -> field.equals(qf)`)

dsmiley replied:
Ah; I see.  So we can have the defaultFieldMatcher in the builder, at least.  Intentionally, many protected methods on UnifiedHighlighter take a field argument because the choice can be field-dependent.  It'd be too awkward to support field-dependent settings in the builder, I think.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -1163,10 +1299,18 @@ public CacheHelper getReaderCacheHelper() {
     PHRASES,
 
-    /** @see UnifiedHighlighter#setHandleMultiTermQuery(boolean) */
+    /**
```
I think these could be javadoc references to the setters on the builder?

dsmiley reviewed (changes requested):
In summary, no setters on UH.  All fields final.  protected get...(field) methods there can simply return the field value.  The builder can compute the flags and any other defaulting logic necessary.
I don't think we need to add tests but some modifications to tests are expected.

apanimesh061 committed:
Addressing comments in PR; Remove setters from UH; Unit tests only use builders

apanimesh061 commented:
#### PR updated

Included:

- Removal of setters and their references from any unit test or class
- Replace UH setters/constructors with the new builder
- Added test UH extensibility
- `flags` are evaluated in the builder

Not included:

- `getFieldMatcher(String field)` is not a getter right now
- No changes related to Javadocs

dsmiley commented on lucene/benchmark/src/java/org/apache/lucene/benchmark/byTask/tasks/SearchTravRetHighlightTask.java:
```diff
@@ -288,17 +283,32 @@ private void reset(IndexSearcher searcher) {
             @Override
-            protected OffsetSource getOffsetSource(String field) {
-              return offsetSource != null ? offsetSource : super.getOffsetSource(field);
+            protected ConcreteBuilder self() {
```
Is overriding `self` actually necessary here?

dsmiley commented on lucene/benchmark/src/java/org/apache/lucene/benchmark/byTask/tasks/SearchTravRetHighlightTask.java:
```diff
@@ -319,4 +329,12 @@ public void withTopDocs(IndexSearcher searcher, Query q, TopDocs hits) throws Ex
   }
+
+  /** This ConcreteBuilder has been created for this class only. */
+  private static class ConcreteBuilder extends UnifiedHighlighter.Builder<ConcreteBuilder> {
```
I doubt we need "ConcreteBuilder" unless we need to add new options.

apanimesh061 replied:
As suggested there is no ConcreteBuilder now. We can override the builder using anonymous classes.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -117,13 +116,6 @@
-  // For analysis, prefer MemoryIndexOffsetStrategy
-  private boolean defaultPassageRelevancyOverSpeed = true;
-
   private int maxLength = DEFAULT_MAX_LENGTH;
```
All fields should be final.

apanimesh061 replied:
All fields have been made final

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -143,88 +135,168 @@
+  private Set<HighlightFlag> flags;
+
+  /** Builder for UnifiedHighlighter. */
+  public abstract static class Builder<T extends Builder<T>> {
```
Why abstract?

dsmiley replied:
I do appreciate the point of the generic type, but I suspect it's value isn't worth the verbosity for the, I believe, common case where a custom Builder type isn't necessary.  It's only necessary when new options are being added.

apanimesh061 replied:
Yes thanks for the feedback here. I'll  work on this builder now.

Sorry this past week I got super busy, so could not get to this.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -143,88 +135,168 @@
+    private int cacheFieldValCharsThreshold = DEFAULT_CACHE_CHARS_THRESHOLD;
+    private Set<HighlightFlag> flags;
+
+    public T withSearcher(IndexSearcher value) {
```
All these "with" methods need Javadocs because users will see these and want to know what they do at a glance.  They won't go looking at protected methods on UnifiedHighlighter for this; that's too much work.  You can simply move these javadocs from where they were and add a javadoc back-reference from those places.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -143,88 +135,168 @@
+    }
 
-  /**
-   * Constructs the highlighter with the given index searcher and analyzer.
```
This good info has no vanished?

apanimesh061 replied:
Moved the info to builder method.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -143,88 +135,168 @@
-  public void setFieldMatcher(Predicate<String> predicate) {
-    this.defaultFieldMatcher = predicate;
+      // Evaluate if WEIGHT_MATCHES can be added as a flag.
+      if (highlightFlags.contains(HighlightFlag.MULTI_TERM_QUERY)
```
I think we should have a `withWeightMatches` as well so that a user can easily opt-out.

dsmiley commented on lucene/highlighter/src/test/org/apache/lucene/search/uhighlight/TestUnifiedHighlighterMTQ.java:
```diff
@@ -117,7 +117,11 @@ public void testWildcards() throws Exception {
 
     IndexSearcher searcher = newSearcher(ir);
-    UnifiedHighlighter highlighter = new UnifiedHighlighter(searcher, indexAnalyzer);
+    UnifiedHighlighter.ConcreteBuilder concreteBuilder =
```
Nitpick: I wouldn't bother declaring an intermediate variable for the builder.  With "fluent" interfaces (what a builder is), you can chain method calls and get the final thing you want -- the UnifiedHighlighter instance itself.

dsmiley commented on lucene/highlighter/src/test/org/apache/lucene/search/uhighlight/TestUnifiedHighlighterMTQ.java:
```diff
@@ -726,54 +752,54 @@ public void testWhichMTQMatched() throws Exception {
         };
 
+    // Now use our own formatter, that also stuffs the
+    // matching term's text into the result:
```
This comment used to immediately precede the formatter definition because it was talking about it.  After your refactoring (which I like), the comment is not in the right place anymore.

dsmiley commented on lucene/highlighter/src/test/org/apache/lucene/search/uhighlight/visibility/TestUnifiedHighlighterExtensibility.java:
```diff
@@ -96,129 +96,14 @@ protected OffsetsEnum createOffsetsEnumFromReader(LeafReader leafReader, int doc
+            .withSearcher(null)
+            .withIndexAnalyzer(new MockAnalyzer(random()))
+            .withNewField(false)
+            .build();
```
Just to be clear, the point of `testUnifiedHighlighterExtensibility` is so that we're aware of potential backwards-compatibility breaks that we might more easily overlook.  It's not a great technique... I've heard of Java build plugins that can alert you to this as well.  Any way, your refactor, I think, obscures this since the subclass is no longer within the method definition.  I think you made this change because you thought we need some typed builder but I don't think we do.  Just have an anonymous implementation of the builder that overrides build() to return an anonymous subclass of the UnifiedHighlighter.

dsmiley reviewed (changes requested):
Lots of great progress!

All in all, I think the most important part of my feedback here is the implications of Builder not needing to be abstract.  I suspect we don't need any typed subclasses of Builder -- if we need to override the UH in order to override one of those protected methods, it can be done in an anonymous subclass of the UH returned from an anonymous subclass of the Builder.

BTW the release will be cut Tuesday so we have another day to make 9.0.

apanimesh061 committed:
Merge branch 'apache:main' into uh-builder-1

apanimesh061 committed:
Addressing comments for PR review on 10/31

apanimesh061 commented:
@dsmiley 
I was finally able to address your comments on my previous PR request. Thank you for your patience.

Added a new PR version with new changes related to the builder. Please take a look whenever you get a chance.

Following commands work:
```
./gradlew preCommit
./gradlew check
```

New version includes:
- Non-abstract builder with no "concreteBuilder"
- More java-doc statements in the UH class
- Modified unit-tests

dsmiley commented on lucene/highlighter/src/test/org/apache/lucene/search/uhighlight/TestUnifiedHighlighterRanking.java:
```diff
@@ -330,8 +328,11 @@ public void testCustomK1() throws Exception {
 
     IndexSearcher searcher = newSearcher(ir);
+
+    UnifiedHighlighter.Builder concreteBuilder =
```
again, maybe `uhBuilder`

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -113,118 +112,239 @@
       new LabelledCharArrayMatcher[0];
 
-  protected final IndexSearcher searcher; // if null, can only use highlightWithoutSearcher
+  protected final IndexSearcher searcher;
```
the comment you removed was useful.  Oh I see you moved it to the builder.  I think internal comments are best kept here.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -113,118 +112,239 @@
+     * Usually required, unless {@link #highlightWithoutSearcher(String, Query, String, int)} is
+     * used, in which case this needs to be null.
+     */
+    public Builder withSearcher(IndexSearcher value) {
```
Given that the searcher is often required (albeit not always), it would be a good candidate for putting in the builder(IndexSearcher) param.  Same with the Analyzer. I've seen this pattern before for builders. This way the caller is forced to think of it, but would otherwise easily forget many of the "with" methods on the builder.

Since there are two primary scenarios -- with and without a searcher, we could have two builder methods: build(IndexSearcher, Analyzer) and buildWithoutSearcher(Analyzer) which thus makes them explicit when reading code.

apanimesh061 replied:
@dsmiley 

Thanks for the feedback.

Would something like this work?

```
UnifiedHighlighter unifiedHighlighter =
        UnifiedHighlighter.builder(searcher, indexAnalyzer)
            .withMaxLength(100)
            .withScorer(new PassageScorer())
            .build();
```
and 

```
UnifiedHighlighter unifiedHighlighter =
        UnifiedHighlighter.builderWithoutSearcher(indexAnalyzer)
            .withMaxLength(100)
            .withScorer(new PassageScorer())
            .build();
```

I'll then remove `withSearcher()` and `withIndexAnalyzer()`.

dsmiley replied:
Yeah; that's exactly what I meant. Thanks.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -113,118 +112,239 @@
-    this.defaultHighlightPhrasesStrictly = highlightPhrasesStrictly;
-  }
+    /**
+     * This method sets the analyzer for the UH object. Required, even if in some circumstances it
```
BTW in Javadoc style, don't start javadocs with "this method" or "this class" -- instead get right to the point and succinctly.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -113,118 +112,239 @@
+      return self();
+    }
+
+    public Builder withMaxLength(int value) {
```
Suggested javadoc:
> The text to be highlight is effectively truncated by this length.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -113,118 +112,239 @@
+     * @return a set of {@link HighlightFlag}s.
+     */
+    protected Set<HighlightFlag> evaluateFlags() {
+      if (Objects.nonNull(flags) && !flags.isEmpty()) {
```
Isn't empty valid?

apanimesh061 replied:
Yeah that makes sense. Will fix.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -143,6 +143,91 @@
+      return self();
+    }
+
+    protected Builder self() {
```
Seems unnecessary now?

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -113,118 +112,239 @@
-   * Returns whether position sensitive queries (e.g. phrases and {@link SpanQuery}ies) should be
-   * highlighted strictly based on query matches (slower) versus any/all occurrences of the
-   * underlying terms. By default it's enabled, but there's no overhead if such queries aren't used.
+   * Constructs the highlighter with the given the {@link Builder}.
```
```suggestion
   * Constructs the highlighter with the given {@link Builder}.
```

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -823,24 +943,13 @@ public void visitLeaf(Query query) {
-  /** Customize the highlighting flags to use by field. */
+  /**
+   * Customize the highlighting flags to use by field. Here the user can either specify the set of
+   * {@link HighlightFlag}s to be applied or use the boolean flags to populate final list of {@link
```
I think this is incorrect now.  The boolean flags only exist in the builder.

apanimesh061 replied:
Will rectify the comment.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -113,118 +112,239 @@
+  private final int maxLength;
 
-  private boolean defaultHighlightPhrasesStrictly = true; // AKA "accuracy" or "query debugging"
+  private final Supplier<BreakIterator> defaultBreakIterator;
```
I suggest we remove the `default` prefix from the fields.  They are all effectively defaults that are often over-rideable.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -1163,23 +1272,16 @@ public CacheHelper getReaderCacheHelper() {
   /** Flags for controlling highlighting behavior. */
   public enum HighlightFlag {
-    /** @see UnifiedHighlighter#setHighlightPhrasesStrictly(boolean) */
+    /** Ref: {@link Builder#withHighlightPhrasesStrictly(boolean)} */
```
What's wrong with `@see` ?

dsmiley reviewed (changes requested):
Awesome progress!
I have one major comment RE builder params but otherwise just minor stuff.

apanimesh061 committed:
Addressing comments for PR review on 12/03

apanimesh061 commented:
@dsmiley 

Updated the PR as per the comments and added the new builder functions.

apanimesh061 committed:
Add javadocs for static builder factory methods

dsmiley commented on lucene/highlighter/src/test/org/apache/lucene/search/uhighlight/TestUnifiedHighlighter.java:
```diff
@@ -137,23 +141,24 @@ static UnifiedHighlighter randomUnifiedHighlighter(
   static UnifiedHighlighter.Builder overriddenBuilderForTests(
-      UnifiedHighlighter.Builder uhBuilder, EnumSet<HighlightFlag> mandatoryFlags) {
-    return new UnifiedHighlighter.Builder() {
+      IndexSearcher searcher,
```
Why would a method ever need both the builder, and components that can be found on the builder?  Just call uhBuilder.getIndexSearcher (or access the field if in scope)

apanimesh061 replied:
I did think about it once but I thought if having a getter for the searcher and analyzer from the Builder class would be good idea. Based on your comment I guess it should be fine. That will simplify a number of helper functions.

apanimesh061 committed:
Addressing comments for PR review on 12/05

dsmiley reviewed (approved):
I think this is done!

Please add an entry to lucene/CHANGES.txt.  I propose the following, under "API Changes":
```
* LUCENE-10197: UnifiedHighlighter now has a mandatory Builder to construct it.  The UH is now immutable.  (your-name, David Smiley)
```

apanimesh061 committed:
Updating CHANGES.txt for LUCENE-10197

apanimesh061 commented:
> I think this is done!
> 
> Please add an entry to lucene/CHANGES.txt. I propose the following, under "API Changes":
> 
> ```
> * LUCENE-10197: UnifiedHighlighter now has a mandatory Builder to construct it.  The UH is now immutable.  (your-name, David Smiley)
> ```

@dsmiley That is great.

I've added the entry to the file under `API Changes` for `Lucene 10.0.0`.

apanimesh061 committed:
Merge branch 'main' into uh-builder-1

apanimesh061 commented:
@dsmiley In order to make this portable to v9.x should we make 2 unit test suites for "with builder" and "w/o builder"? It will just be easier to remove the non-builder files and rename the builder based tests in v10.x.

dsmiley commented:
Let's definitely *not* double up the tests in the name of test coverage.  It's not worth the maintenance overhead.  As it is, I think there's some other double-up test for something that didn't get reverted.  The deprecated methods (there for mutability) can be un-tested.  We can use the builder everywhere henceforth.

apanimesh061 committed:
Attempt at making UH with builder back-portable to v9.x

apanimesh061 commented:
@dsmiley I've made an attempt at making this back-portable as suggested.

- `@Deprecated` the setters and constructor
- Have not added any tests for the non-builder case
- Existing tests are working

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -360,14 +451,104 @@ public UnifiedHighlighter(Builder builder) {
+   * @return {@link HighlightFlag}s.
+   */
+  protected Set<HighlightFlag> evaluateFlags(Builder uhBuilder) {
+    if (Objects.nonNull(flags)) {
```
Objects.nonNull is useful when needing to reference as a lambda but otherwise is an awkward way to simply say != null

apanimesh061 replied:
Yeah that is a good point. Will replace it with `!= null`.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -383,6 +564,15 @@ public UnifiedHighlighter(Builder builder) {
 
+  /** Returns the {@link HighlightFlag}s applicable for the current UH instance. */
+  protected Set<HighlightFlag> getFlags(String field) {
+    if (initByBuilder) {
```
instead of having `initByBuilder`, can we just have a null-able flags field that can only be non-null if the builder was used?

apanimesh061 replied:
Fixed.

dsmiley commented on lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java:
```diff
@@ -97,6 +97,23 @@
 
   public static final int DEFAULT_CACHE_CHARS_THRESHOLD = 524288; // ~ 1 MB (2 byte chars)
 
+  public static final boolean DEFAULT_ENABLE_MULTI_TERM_QUERY = true;
```
Let's make these private so that we needn't keep them longer term.  They are only useful for this period of time in which we have both a setter & builder.  A little comment here stating this would be useful so we know to remove them eventually.

apanimesh061 replied:
Fixed

apanimesh061 committed:
Addressing PR comments from 12/19

apanimesh061 commented:
@dsmiley I've made the suggested modifications.

dsmiley committed:
Merge branch 'main' into uh-builder-1
